doc/compiler/optimizer/LocalOptimizationsSummary.md

```diff
@@ -0,0 +1,103 @@
+-->
+
+# Local Optimizations
+
+Local optimizations operate on extended basic blocks at a time.
+
+## Tree Simplification
+Tree Simplification performs simple IL tree transformations like constant
+folding, strength reduction, canonicalizing expressions (e.g. an `add` with a
+constant child will have the constant as the second child), branch/switch
+simplification etc. Float/double simplification support needs to be improved
+(call C libraries to compute the correct answer as per the Java spec).
+
+## Local Common Subexpression Elimination (CSE)/Copy Propagation
+Local CSE/Copy Propagation commons identical expressions within a block and
+performs removal of identical checks along with traditional commoning. Copy
+Propagation is done simultaneously (The right hand side (RHS) of a store is
+evaluated and the resulting register is used at each use of the stored value).
+Commoning is done on syntactic equivalence.
+
+## Local Value Propagation
+Local [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md)
+performs constant, type and relational propagation within a block.
+It propagates constants and types based on assignments and performs removal
+of subsumed checks and compares. It also performs devirtualization
+of calls, checkcast and array store check elimination. It does not do value
+numbering, instead it works on the assumption that every expression has a
+unique value number (improved effectiveness if expressions having same value
+are commoned by a prior local CSE pass).
+
+## Local Dead Store Elimination
+Local Dead Store Elimination eliminates stores to locals/fields/statics (if
+there are two stores to the same memory without any intervening use in the
+same block).
+
+## Local Reordering of Expressions
+Local Reordering of Expressions reduces live ranges (register pressure) by
+moving evaluations around within a block.
+
+## Array Length Simplification
+In OpenJ9, array lengths are stored in number of bytes. So bound checks require
+a shift right (depending on the array stride) on the array length so that
+compare can be done with array index. This optimization attempts to eliminate
+`shr` wherever possible (instead comparing the index left shifted based on
+array stride with the array length in bytes) based on when it can be proven
+that it is legal to do so. Note that shifting left may cause wrap-arounds (if
+the value is close to MAXINT for example) and so this optimization is done
+for constant indices or indices marked as safe to left shift by value propagation.
+Shifting left is better as it is already required in most cases for the address
+computation for the array access.
+
+## Compaction of Null Explicit Null Checks
+Compaction of Null Explicit Null Checks minimize the cases where an explicit null
+check needs to be performed (null checks are implicit, i.e. no code for doing null
+check, instead we use hardware trap on IA32). Sometimes (say as a result of dead
```

Comment:
```suggestion
check needs to be performed (null checks are implicit, i.e. by default on IA32, the JIT does not
emit code to perform a null check, instead we use a hardware trap). Sometimes (e.g. as a result of dead
```

Reply:
Do we have implicit null checks only on IA32?

Reply:
No, we have them on X86 (32 and 64 bitness). IA32 was just the quaint way of referring to the platform when "64-bit was new".